File: examples/pay-lion.mjs
// LION x402 keyless reference client (EIP-3009 / EIP-712 -> Payment-Signature)
//
// Pays an HTTP 402 LION route in USDC on Base with any funded wallet.
// No API key, no account: the signed transfer authorization IS the payment.
//
// Usage:
//   npm install ethers
//   LION_PK=0x<base-wallet-private-key> node --max-http-header-size=131072 pay-lion.mjs \
//     "https://lionx402.com/api/x402/cpg-product-intel-json?barcode=5449000000996"
//
// The private key is read from the LION_PK env var so it never appears in shell history.
// Use a dedicated low-balance wallet.

import { ethers } from "ethers";
import crypto from "node:crypto";

const b64u = (buf) => Buffer.from(buf).toString("base64url");
const RESOURCE = process.argv[2];
let PK = (process.env.LION_PK || "").trim();
if (PK && !PK.startsWith("0x")) PK = "0x" + PK;
if (!PK) { console.error("ERROR: set LION_PK to your funded Base wallet private key"); process.exit(1); }
if (!RESOURCE) { console.error("ERROR: pass the paid resource URL as argument 1"); process.exit(1); }

// Pick the keyless xpay accept (x402 v1, exact scheme on Base).
function pickAccept(body) {
  const accepts = body.accepts || [];
  return accepts.find(a => a.scheme === "exact" && (a.extra?.facilitatorKind === "xpay" || a.x402Version === 1)) || accepts[0];
}

async function buildEnvelope(accept, wallet) {
  const chainId = 8453;
  const asset = accept.asset;
  const value = String(accept.amount || accept.maxAmountRequired);
  const to = accept.payTo;
  const from = wallet.address;
  const now = Math.floor(Date.now() / 1000);
  const authorization = {
    from, to, value,
    validAfter: now - 60,
    validBefore: now + 600,
    nonce: "0x" + crypto.randomBytes(32).toString("hex"), // unique per signature, forever
  };

  const domain = { name: accept.extra?.name || "USD Coin", version: accept.extra?.version || "2", chainId, verifyingContract: asset };
  const types = { TransferWithAuthorization: [
    { name: "from", type: "address" }, { name: "to", type: "address" },
    { name: "value", type: "uint256" }, { name: "validAfter", type: "uint256" },
    { name: "validBefore", type: "uint256" }, { name: "nonce", type: "bytes32" } ] };

  const signature = await wallet.signTypedData(domain, types, authorization);

  // The FULL x402 envelope (not just {signature, authorization}) is required.
  return {
    x402Version: accept.x402Version || 1,
    scheme: accept.scheme || "exact",
    network: accept.network || "base",
    asset,
    payTo: to,
    resource: accept.resource,
    maxAmountRequired: value,
    amount: value,
    payload: { signature, authorization },
  };
}

const wallet = new ethers.Wallet(PK);
console.error("payer:", wallet.address);

const challenge = await (await fetch(RESOURCE, { headers: { accept: "application/json" } })).json();
if (!challenge.accepts) { console.log(JSON.stringify({ note: "no 402 (free/already paid)", challenge }, null, 2)); process.exit(0); }

const accept = pickAccept(challenge);
console.error("paying:", Number(accept.amount || accept.maxAmountRequired) / 1e6, "USDC ->", accept.payTo);

const envelope = await buildEnvelope(accept, wallet);
// Header value = base64url(JSON.stringify(envelope)). Server base64url-decodes then JSON.parses.
const header = b64u(JSON.stringify(envelope));

const res = await fetch(RESOURCE, { headers: { accept: "application/json", "Payment-Signature": header } });
const text = await res.text();
let body; try { body = JSON.parse(text); } catch { body = text; }
console.log("HTTP", res.status);
console.log(JSON.stringify(body, null, 2));

File: examples/pay_lion.py
#!/usr/bin/env python3
"""LION x402 keyless reference client (Python).

Pays an HTTP 402 LION route in USDC on Base with any funded wallet. No API key,
no account: the signed EIP-3009 transfer authorization IS the payment.

    pip install -r requirements.txt      # just eth-account
    LION_PK=0x<base-wallet-private-key> python3 pay_lion.py \
        "https://lionx402.com/api/x402/cpg-product-intel-json?barcode=5449000000996"

The private key is read from LION_PK so it never lands in shell history. Use a
dedicated, low-balance wallet funded with a little USDC on Base.

Signing is byte-for-byte identical to the proven JS reference (examples/pay-lion.mjs);
this client only differs in language, not in the bytes it puts on the wire.
"""
import base64
import json
import os
import secrets
import sys
import time
import urllib.error
import urllib.request

from eth_account import Account
from eth_account.messages import encode_typed_data

CHAIN_ID = 8453  # Base


def b64u(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode().rstrip("=")


def http(url: str, headers=None):
    """GET returning (status, text); 402/4xx bodies are returned, not raised."""
    hdrs = {"accept": "application/json", "user-agent": "lion-x402-example/1.0"}
    hdrs.update(headers or {})
    req = urllib.request.Request(url, headers=hdrs)
    try:
        r = urllib.request.urlopen(req)
        return r.status, r.read().decode()
    except urllib.error.HTTPError as e:
        return e.code, e.read().decode()


def pick_accept(body):
    accepts = body.get("accepts") or []
    for a in accepts:
        extra = a.get("extra") or {}
        if a.get("scheme") == "exact" and (extra.get("facilitatorKind") == "xpay" or a.get("x402Version") == 1):
            return a
    return accepts[0] if accepts else None


def main():
    if len(sys.argv) < 2:
        sys.exit("ERROR: pass the paid resource URL as argument 1")
    resource = sys.argv[1]
    pk = (os.environ.get("LION_PK") or "").strip()
    if pk and not pk.startswith("0x"):
        pk = "0x" + pk
    if not pk:
        sys.exit("ERROR: set LION_PK to your funded Base wallet private key")

    acct = Account.from_key(pk)
    print("payer:", acct.address, file=sys.stderr)

    status, text = http(resource)
    challenge = json.loads(text)
    if not challenge.get("accepts"):
        print(json.dumps({"note": "no 402 (free/already paid)", "challenge": challenge}, indent=2))
        return

    accept = pick_accept(challenge)
    value = str(accept.get("amount") or accept.get("maxAmountRequired"))
    asset = accept["asset"]
    pay_to = accept["payTo"]
    print(f"paying: {int(value) / 1e6} USDC -> {pay_to}", file=sys.stderr)

    now = int(time.time())
    nonce_hex = secrets.token_hex(32)  # unique per signature, forever
    authorization = {
        "from": acct.address,
        "to": pay_to,
        "value": int(value),
        "validAfter": now - 60,
        "validBefore": now + 600,
        "nonce": bytes.fromhex(nonce_hex),
    }
    extra = accept.get("extra") or {}
    full_message = {
        "types": {
            "EIP712Domain": [
                {"name": "name", "type": "string"},
                {"name": "version", "type": "string"},
                {"name": "chainId", "type": "uint256"},
                {"name": "verifyingContract", "type": "address"},
            ],
            "TransferWithAuthorization": [
                {"name": "from", "type": "address"},
                {"name": "to", "type": "address"},
                {"name": "value", "type": "uint256"},
                {"name": "validAfter", "type": "uint256"},
                {"name": "validBefore", "type": "uint256"},
                {"name": "nonce", "type": "bytes32"},
            ],
        },
        "domain": {
            "name": extra.get("name") or "USD Coin",
            "version": extra.get("version") or "2",
            "chainId": CHAIN_ID,
            "verifyingContract": asset,
        },
        "primaryType": "TransferWithAuthorization",
        "message": authorization,
    }
    signature = acct.sign_message(encode_typed_data(full_message=full_message)).signature.hex()
    if not signature.startswith("0x"):
        signature = "0x" + signature

    # The FULL x402 envelope (not just {signature, authorization}) is required.
    envelope = {
        "x402Version": accept.get("x402Version") or 1,
        "scheme": accept.get("scheme") or "exact",
        "network": accept.get("network") or "base",
        "asset": asset,
        "payTo": pay_to,
        "resource": accept.get("resource"),
        "maxAmountRequired": value,
        "amount": value,
        "payload": {
            "signature": signature,
            "authorization": {
                "from": acct.address,
                "to": pay_to,
                "value": value,
                "validAfter": authorization["validAfter"],
                "validBefore": authorization["validBefore"],
                "nonce": "0x" + nonce_hex,
            },
        },
    }
    # Header value = base64url(JSON). Server base64url-decodes then JSON.parses.
    header = b64u(json.dumps(envelope, separators=(",", ":")).encode())
    status, text = http(resource, {"accept": "application/json", "Payment-Signature": header})
    print("HTTP", status)
    try:
        print(json.dumps(json.loads(text), indent=2))
    except json.JSONDecodeError:
        print(text)


if __name__ == "__main__":
    main()
